pkg/chain/ethereum/ethutil/rate_limiter_test.go

```diff
@@ -0,0 +1,350 @@
+package ethutil
+
+import (
+	"context"
+	"github.com/ethereum/go-ethereum"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/core/types"
+	"math/big"
+	"strings"
+	"sync"
+	"testing"
+	"time"
+)
+
+func TestRateLimiter(t *testing.T) {
+	requestsPerSecondLimit := 500
+	concurrencyLimit := 5
+	acquirePermitTimeout := time.Minute
+	requests := 500
+	requestDuration := 10 * time.Millisecond
```

Comment:
Shouldn't we have a separate test for rate-limiting and concurrency level?

Even if I comment out `RequestsPerSecondLimit` in `RateLimiterConfig` this test is still passing because of how `concurrencyLimit` and `requestDuration` are configured. We can have at most `1000 / 10 * 5` requests per second.

Reply:
Alternatively, I think we can try to generate random request durations in some range.

Reply:
I can add some separate tests but in general this is how it works. Limiting only the `RequestsPerSecondLimit` always have an impact on concurrency and using the `ConcurrencyLimit` alone will impact the requests per second rate as well.

Reply:
Yeah, I do understand they influence each other but just from the test perspective, I'd like to have a code to ensure they both work the way we expect it. Right now, we don't know whether rate-limiter works because concurrency limit is taking the entire hit in this test.

Reply:
Ok I will rework those tests a bit.

Reply:
Done.